### ISSUE TITLE
Listen:  grp's "playlist-3" and "playlist-4" do not play

### ISSUE THREAD

**Issue by Brahmanathaswami** [Himalayan-Academy/Siva-Siva-App#287]
They return:

<img width="414" alt="image" src="https://user-images.githubusercontent.com/3675722/113806502-17cb6f80-96fe-11eb-89e9-a25122351599.png">

**Comment by Brahmanathaswami** [comment]
git: 
There come up on "nightly"

compare nightly 
       behavior_Listen 

and listen-5
        behavior_Listen
        
        it should be  easy to  fix

**Comment by soapdog** [comment]
This appears to be related to the arrays having the wrong elements in them, as can be seen in the screenshot below:

![com livecode palette autocomplete completions 2021-04-20 11-08-56](https://user-images.githubusercontent.com/23247/115379175-2f5c2b00-a1c9-11eb-80c1-935976b80731.png)

**Comment by soapdog** [comment]
My suspicion is that the code that fetches and caches data is buggy, or that the server part is answering with bad content.

**Comment by soapdog** [comment]
Data is loaded from a json file at `modules/listen/collection.json`. This json doesn't contain the fields that the code is attempt to access for some collections, as can be seen below:

```json
{
        "playListTitle": "Bodhinatha's\nLatest",
        "playListImg": "bodhinatha-latest_330.jpg",
        "playlistClass": "latest",
        "recordType": "latest",
        "recordCount": "30",
        "data_query": {
          "media_type": "inspiredtalk",
          "author": "Satguru Bodhinatha Veylanswami",
          "approved_for_public": "YES"
        }
      },
      {
        "playListTitle": "Classical\nArtists",
        "playListImg": "professional_rajam-singer_330.jpg",
        "playlistClass": "cached",
        "data_query": {
          "media_type": "audio/song",
          "tags": "Professional",
          "approved_for_public": "YES"
        }
      },
      {
        "playListTitle": "Popular Tunes",
        "playListImg": "popular-tunes-lady-headset_shutterstock_330.jpg",
        "playlistClass": "cached",
        "data_query": {
          "media_type": "audio/song",
          "tags": "pop",
          "approved_for_public": "YES"
        }
      }
    ]
  },
```

The first item in the snippet above has `recordType` and `recordCount`, but the two items next to it don't.

**Comment by soapdog** [comment]
Then the code proceeds to fetch data from the bundled jnanam database. I've altered `fetchMediaItems` function in `lib_SivaSivaMedia` because it was using _string_ functions to search for _numbers_. I believe that this change might be leading to this problem as the _string_ comparison being used requires the match to be _exact_ and not _partial_.

Changed code, notice the `dbLike` command requiring exact matches:

![image](https://user-images.githubusercontent.com/23247/115386248-0dff3d00-a1d1-11eb-9797-ac03a23fff30.png)

**Comment by soapdog** [comment]
Yes, that is the source of the problem. `fetchMediaItems` needs to use partial match or it won't be able to match tags.

A shot below is showing and example result with partial matches, notice the tags field.

![image](https://user-images.githubusercontent.com/23247/115386980-f96f7480-a1d1-11eb-84a8-a8fe4e1e987e.png)

Using partial matches in that code will lead to false matches for some cases as some words are part of other words. For example, if we change it to a partial match, searching for `Guru` will also match `Gurudeva` and `Satgurus`, which are different tags.

**Comment by soapdog** [comment]
In the original bug submission, @Brahmanathaswami wrote

> git:
> There come up on "nightly"
> 
> compare nightly
> behavior_Listen
> 
> and listen-5
> behavior_Listen
> 
> ```
>     it should be  easy to  fix
> ```

I just want to highlight that these are not easy fixes and they were not related to `behavior_Listen`. The code in this app has some _extreme tight couplings_ and is very convoluted with handlers that are overloaded with special behaviours and _side-effects depending on which parameters are passed_ to them. Solving bugs in this app require one to explore a lot of behaviours and libraries.

----

In summary, there was a bug in `lib_SivaSivaMedia` that was fixed a month ago. [It was using _string_ comparison SQL code to find _numbers_](https://github.com/Himalayan-Academy/Siva-Siva-App/commit/8353e1d1b71f4b9df43b6127d8317d0d98933c44). This meant that searching for `item_id = 11` would return `11, 111, 1011` and any number that had two ones next to one another. 

When the fix was written, the `dbLike` clause for string comparison was made to be an exact match, so that searching for `guru` would not return partial matches such as `satguru` and `gurudeva`. 

This in turn broke `playlist_FetchData` in `behavior_Listen` because it is searching for tags and tags are a _comma-delimited list_, so the only way to search them with "LIKE" is to use a _partial match_.

The commit to fix this issue, reintroduces partial matches with a special case just for _tags_ fields. This will necessarily introduce a bug in which words that have the same letters are other words will cause a match.

**Comment by soapdog** [comment]
So this is fixed with caveats.

**Comment by Brahmanathaswami** [comment]
I did a Git pull on listen-5.

git log --> `fetchMediaItems` -->   This closes #287

But still need to do something to fix this?

<img width="433" alt="image" src="https://user-images.githubusercontent.com/3675722/115415070-4c293c00-a192-11eb-993d-1426b26191f3.png">

**Comment by soapdog** [comment]
That bug happened because `fetchMediaItems` uses **EXACT MATCHES** for all fields except _tags_ and whoever created the JSON didn't fill the complete information for the `media_type` of that collection.

![Image 2021-04-21 10-28-09](https://user-images.githubusercontent.com/23247/115537588-deb10480-a292-11eb-84e8-b908e90b6f59.png)

It was working before because the code for `fetchMediaItems` was using partial matches for everything, which is wrong.

Fixing the `collection.json` entry to be `audio/inspiredtalk` made it work.

I see that you're relying on a partial match for the _recent additions_, that can be done by changing it from `audio` to `audio%` and being explicit about how the matching should work.

I just spent some time checking the source to learn that _recent additions_ are coming from the server even though they are placed in `collection.json` with properties similar to the ones that are being searched in the local db. That is confusing.
